File: JavaScript/examples/1-simple.js
'use strict';

const EventEmitter = function() {
  this.events = {/* hash of array of function */};
};

EventEmitter.prototype.on = function(name, fn) {
  const event = this.events[name] || [];
  this.events[name] = event;
  event.push(fn);
};

EventEmitter.prototype.emit = function(name, data) {
  const event = this.events[name];
  if (event) event.forEach(fn => fn(data));
};

module.exports = EventEmitter;

File: Haskell/1-simple.hs
module EventEmitter where

import Data.Map.Strict as M
import Data.Maybe

type EventEmitter a f = Map a [f]

on :: Ord a => a -> f  -> EventEmitter a f  -> EventEmitter a f
on name fn = alter (Just . maybe [fn] (fn:)) name

emit :: Ord a => a -> b -> EventEmitter a (b -> c) -> [c]
emit name dat = fmap ($ dat) . fromMaybe [] . M.lookup name

newEventEmitter :: EventEmitter a f
newEventEmitter = empty
